map/local_ads_manager.cpp

```diff
@@ -164,66 +150,63 @@ void LocalAdsManager::UpdateViewport(ScreenBase const & screen)
 void LocalAdsManager::ThreadRoutine()
 {
   std::string const campaignFile = GetPath(kCampaignFile);
-  std::string const expirationFile = GetPath(kExpirationFile);
 
-  // Read persistence data (expiration file must be read first).
-  ReadExpirationFile(expirationFile);
+  // Read persistence data.
   ReadCampaignFile(campaignFile);
+  Invalidate();
 
   std::vector<Request> campaignMwms;
   while (WaitForRequest(campaignMwms))
   {
-    try
+    for (auto const & mwm : campaignMwms)
     {
-      FilesContainerW campaignsContainer(campaignFile, FileWriter::OP_WRITE_EXISTING);
-      FilesContainerW expirationContainer(expirationFile, FileWriter::OP_WRITE_EXISTING);
-      for (auto const & mwm : campaignMwms)
+      if (!mwm.first.IsAlive())
+        continue;
+
+      std::string const countryName = mwm.first.GetInfo()->GetCountryName();
+      if (mwm.second == RequestType::Download)
       {
-        if (!mwm.first.IsAlive())
+        // Download campaign data from server.
+        CampaignInfo info;
+        info.m_data = DownloadCampaign(mwm.first);
+        if (info.m_data.empty())
           continue;
+        info.m_creationTimestamp = std::chrono::steady_clock::now();
```

Comment:
I suggest you to create a helper static method, say ```Now()``` that will be a wrapper for std chrono steady_clock now.